draft-specifications/Formats.md

```diff
@@ -0,0 +1,248 @@
+  754 64-bit floating point number, rounding ties to even.
+- _wire-float64_: `D` followed by the corresponding 8 bytes of an IEEE 754
+  64-bit floating point number.
+
+## String
+
+A sequence of [Unicode](https://www.unicode.org/standard/standard.html) code
+points excluding surrogates (U+D800-U+DFFF).
+
+> Example: `"twine"` corresponds to `5"twine`.
+
+- _presentation-string_: `"` _presentation-character_ * `"`
+- _presentation-character_: _any printable ASCII character except `"` or `\`_ ;
+  note that spaces are printable
+- _wire-string_: _length_ `"` _bytes_
+- _length_: _integer-digits_ ; The number of bytes in _bytes_ as ASCII decimal
+  digits.
+- _bytes_: _byte_* ; The bytes of the string in UTF-8 encoding.
+
+We do not attempt to capture strings with embedded quotes or non-ASCII Unicode
+characters in the Presentation Format, but all Unicode strings excluding
+surrogates (U+D800-U+DFFF) are expressible in the Wire Format.
+
+## Selector
+
+A sequence of Unicode code points excluding surrogates (U+D800-U+DFFF).
+
+> Example: `'fleur-de-lis` corresponds to `12'fleur-de-lis`.
+
+- _presentation-selector_: `'` _name_
+- _name_: _alpha_ ( _alpha_ / _digit_ / `-` / `:` )*
+- _alpha_: ( `a` - `z` ) / ( `A` - `Z` )
+- _digit_: `0` - `9`
+- _wire-selector_: _length_ `'` _bytes_
+- _bytes_: _byte_* ; The bytes of the selector in UTF-8 encoding.
+
+We do not attempt to capture selectors with arbitrary Unicode characters in the
+Presentation Format, but all Unicode strings excluding surrogates
+(U+D800-U+DFFF) are expressible in the Wire Format.
+
+## ByteArray
+
+An array of 8-bit bytes.
+
+> Example: `:b0b5c0ffeefacade` corresponds to the bytes of `8`, `:`, b0, b5,
+> c0, ff, ee, fa, ca, de in ASCII and hexadecimal.
+
+- _presentation-byte-array_: `:` _hex_
+- _hex_: ( _hex-digit_ _hex-digit_ )*
+- _hex-digit_: _digit_ / ( `a` - `f` ) ; Corresponding to a _byte_ of _bytes_.
+- _wire-byte-array_: _length_ `:` _bytes_ ; The number of bytes in _bytes_
+  and the _bytes_.
+
+We do not attempt to capture byte arrays with non-space ASCII characters in the
+Presentation Format, but all byte arrays are expressible in the Wire Format.
+
+## Struct
+
+> The name "struct" is tentative.
+> https://github.com/ocapn/ocapn/pull/125
+
+A collection of (field name, value) pairs.
+
+In the Presentation Format and Wire Format, no field name may appear more than
+once.
+
+In the Wire Format, fields must appear in strictly-ascending bytewise order.
+
+> Examples:
+> - `{ a: 10, b: 2 }` corresponds to `{1"a10+1"b2+}`.
+> - `{ "a": 10, "b": 2 }` corresponds to `{1"a10+1"b2+}`.
+> - `{ 'b: 2, 'a: 10 }` corresponds to `{1'a10+1'b2+}`.
+
+- _presentation-struct_: `{` ( _presentation-field_ ( `,` _presentation-field_ )* )? `}`
+- _presentation-field_: _presentation-key_ `:` _presentation-value_
+- _presentation-key_: _presentation-field-name_ / _presentation-value_
+- _presentation-field-name_: _name_ ; Corresponding to a string.
```

Comment:
Assuming we come to agreement that this should remain "Corresponding to a string", the String presentation format has an initial and terminal double quote, which explains the second example line above. I would like to allow the first example line, even at the cost of defining a presentation-identifier or something.

Allowing the third example line, for a presentation format for an interoperable protocol seems weird, obscure, and unnecessary. But until I understand better what purpose you have in mind for the Presentation Format, I may be confused about this as well.

Reply:
Nevermind my question about the first example line. I see you dealt with it below. We therefore need to define presentation-identifier or something.